azurerm/internal/services/compute/virtual_machine.go

```diff
@@ -411,3 +416,407 @@ func flattenVirtualMachineOSDisk(ctx context.Context, disksClient *compute.Disks
+					ValidateFunc: validation.StringInSlice([]string{
+						// note: OS Disks don't support Ultra SSDs
+						string(compute.StorageAccountTypesPremiumLRS),
+						string(compute.StorageAccountTypesStandardLRS),
+						string(compute.StorageAccountTypesStandardSSDLRS),
+					}, false),
+				},
+
+				"write_accelerator_enabled": {
+					Type:     schema.TypeBool,
+					Optional: true,
+					Default:  false,
+				},
+			},
+		},
+	}
+}
+
+func expandVirtualMachineDataDisks(d *schema.ResourceData, meta interface{}) (*[]compute.DataDisk, error) {
+	input, ok := d.GetOk("data_disks")
+	if !ok {
+		return &[]compute.DataDisk{}, nil
+	}
+
+	var err error
+	dataDisksRaw := input.([]interface{})
+
+	result := make([]compute.DataDisk, 0)
+	dataDisks := dataDisksRaw[0].(map[string]interface{})
+	if newDisksRaw, ok := dataDisks["local"]; ok {
+		var newDisks []compute.DataDisk
+		if d.IsNewResource() {
+			newDisks = expandVirtualMachineNewDataDisksForCreate(newDisksRaw)
+		} else {
+			newDisks, err = expandVirtualMachineLocalDataDisksForUpdate(newDisksRaw, d, meta)
+			if err != nil {
+				return nil, err
+			}
+		}
+		result = append(result, newDisks...)
+	}
+
+	if existingDisksRaw, ok := dataDisks["existing"]; ok {
+		existingDisks, err := expandVirtualMachineExistingDataDisksForCreate(existingDisksRaw, d, meta)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, existingDisks...)
+	}
+
+	return &result, nil
+}
+
+func expandVirtualMachineNewDataDisksForCreate(input interface{}) []compute.DataDisk {
+	if input == nil || len(input.(*schema.Set).List()) == 0 {
+		return []compute.DataDisk{}
+	}
+
+	var dataDisks []compute.DataDisk
+
+	for _, v := range input.(*schema.Set).List() {
+		disk := v.(map[string]interface{})
+		dataDisk := compute.DataDisk{
+			Caching:      compute.CachingTypes(disk["caching"].(string)),
+			CreateOption: compute.DiskCreateOptionTypesEmpty,
+			DiskSizeGB:   utils.Int32(int32(disk["disk_size_gb"].(int))),
+			Lun:          utils.Int32(int32(disk["lun"].(int))),
+			Name:         utils.String(disk["name"].(string)),
+			ManagedDisk: &compute.ManagedDiskParameters{
+				StorageAccountType: compute.StorageAccountTypes(disk["storage_account_type"].(string)),
+			},
+		}
+		if diskEncryptionSetID, ok := disk["disk_encryption_set_id"].(string); ok && diskEncryptionSetID != "" {
+			dataDisk.ManagedDisk.DiskEncryptionSet = &compute.DiskEncryptionSetParameters{
+				ID: utils.String(diskEncryptionSetID),
+			}
+		}
+		dataDisks = append(dataDisks, dataDisk)
+	}
+
+	return dataDisks
+}
+
+func expandVirtualMachineLocalDataDisksForUpdate(input interface{}, d *schema.ResourceData, meta interface{}) ([]compute.DataDisk, error) {
+	if input == nil || len(input.(*schema.Set).List()) == 0 {
+		return []compute.DataDisk{}, nil
+	}
+
+	diskClient := meta.(*clients.Client).Compute.DisksClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
```

Comment:
we should pass the context (and client) into this method - else we'll have 2 contexts firing (the Update, which is valid) and a Read (5m) at staggered times